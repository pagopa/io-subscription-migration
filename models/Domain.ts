import {
  EmailString,
  NonEmptyString,
  OrganizationFiscalCode
} from "@pagopa/ts-commons/lib/strings";
import * as t from "io-ts";
import {
  ApimSubscriptionResponse,
  ApimUserResponse
} from "./DomainApimResponse";

export type OwnerData = t.TypeOf<typeof OwnerData>;
export const OwnerData = t.intersection(
  [ApimSubscriptionResponse, ApimUserResponse],
  "OwnerData"
);

export type MigrationRowDataTable = t.TypeOf<typeof MigrationRowDataTable>;
export const MigrationRowDataTable = t.interface({
  organizationFiscalCode: OrganizationFiscalCode,
  serviceName: t.string,
  serviceVersion: t.number,
  sourceEmail: EmailString,
  sourceId: NonEmptyString,
  sourceName: NonEmptyString,
  sourceSurname: NonEmptyString,
  subscriptionId: NonEmptyString
});
