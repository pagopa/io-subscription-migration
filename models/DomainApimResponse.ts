import {
  EmailString,
  NonEmptyString,
  OrganizationFiscalCode
} from "@pagopa/ts-commons/lib/strings";
import * as E from "fp-ts/lib/Either";
import { withDefault } from "@pagopa/ts-commons/lib/types";
import * as t from "io-ts";
import { identity, pipe } from "fp-ts/lib/function";

/**
 * Check data of a ApimUserResponse to parse if it's an Organization account.
 * By convention, we decided Organization accounts store the OrganizationFiscalCode as note
 * Thus, by parsing notes we can infer it's an Organization
 *
 * @param u
 * @returns
 */
const isOrganizationUser = (u: RawApimUserResponse): boolean =>
  pipe(
    u,
    E.fromPredicate(
      () => u.note !== undefined,
      () => E.toError
    ),
    E.map(() => pipe(u.note.trim(), OrganizationFiscalCode.decode, E.isRight)),
    E.mapLeft(() => false),
    E.fold(() => false, identity)
  );

export type RawApimUserResponse = t.TypeOf<typeof RawApimUserResponse>;
export const RawApimUserResponse = t.interface({
  email: EmailString,
  firstName: NonEmptyString,
  id: NonEmptyString,
  lastName: NonEmptyString,
  note: withDefault(t.string, "")
});

export type ApimDelegateUserResponse = t.TypeOf<
  typeof ApimDelegateUserResponse
>;
export const ApimDelegateUserResponse = new t.Type<
  RawApimUserResponse & { readonly kind: "delegate" },
  RawApimUserResponse,
  unknown
>(
  "ApimDelegateUserResponse",
  (u): u is RawApimUserResponse & { readonly kind: "delegate" } =>
    RawApimUserResponse.is(u) && !isOrganizationUser(u),
  (u, c) => {
    const maybeRaw = RawApimUserResponse.decode(u);
    if (E.isLeft(maybeRaw)) {
      return t.failure(u, c);
    }
    return !isOrganizationUser(maybeRaw.right)
      ? t.success({
          ...maybeRaw.right,
          kind: "delegate" as const
        })
      : t.failure(u, c);
  },
  RawApimUserResponse.encode
);

export type ApimOrganizationUserResponse = t.TypeOf<
  typeof ApimOrganizationUserResponse
>;
export const ApimOrganizationUserResponse = new t.Type<
  RawApimUserResponse & { readonly kind: "organization" },
  RawApimUserResponse,
  unknown
>(
  "ApimOrganizationUserResponse",
  (u): u is RawApimUserResponse & { readonly kind: "organization" } =>
    RawApimUserResponse.is(u) && isOrganizationUser(u),
  (u, c) => {
    const maybeRaw = RawApimUserResponse.decode(u);
    if (E.isLeft(maybeRaw)) {
      return t.failure(u, c);
    }
    return isOrganizationUser(maybeRaw.right)
      ? t.success({
          ...maybeRaw.right,
          kind: "organization" as const
        })
      : t.failure(u, c);
  },
  RawApimUserResponse.encode
);

export type ApimUserResponse = t.TypeOf<typeof ApimUserResponse>;
export const ApimUserResponse = t.union([
  ApimDelegateUserResponse,
  ApimOrganizationUserResponse
]);

const ApimSubscriptionResponseR = t.interface({
  ownerId: NonEmptyString,
  subscriptionId: NonEmptyString
});
const ApimSubscriptionResponseO = t.partial({});

export const ApimSubscriptionResponse = t.exact(
  t.intersection(
    [ApimSubscriptionResponseR, ApimSubscriptionResponseO],
    "ApimSubscriptionResponse"
  )
);
export type ApimSubscriptionResponse = t.TypeOf<
  typeof ApimSubscriptionResponseR
>;
