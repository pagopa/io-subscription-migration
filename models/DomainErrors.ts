import { ErrorResponse as ApimErrorResponse } from "@azure/arm-apimanagement";

import { DatabaseError } from "pg";

export interface IDbError {
  readonly kind: "dberror";
  readonly message?: string;
}

export interface IApimSubError {
  readonly kind: "apimsuberror";
  readonly message: string;
}

export interface IApimUserError {
  readonly kind: "apimusererror";
  readonly message: string;
}

export const toApimSubError = (message: string): IApimSubError => ({
  kind: "apimsuberror",
  message
});

export const toApimUserError = (message: string): IApimUserError => ({
  kind: "apimusererror",
  message
});

export const toPostgreSQLError = (message: string): IDbError => ({
  kind: "dberror",
  message
});

export type DomainError = IDbError | IApimSubError | IApimUserError;

export const toString = (err: DomainError): string =>
  `${err.kind}|${err.message}`;

export const toPostgreSQLErrorMessage = ({
  code,
  message,
  detail
}: DatabaseError): string => {
  switch (code) {
    case "23505":
      return `Duplicate Primary Key|${code}|${message}|${detail}`;
    case "42P00":
      return `Table not found|${code}|${message}|${detail}`;
    default:
      return `DB Generic Error|${code ||
        "no-code-returned"}|${message}|${detail}`;
  }
};

export const toApimSubErrorMessage = ({
  statusCode: code,
  message
}: ApimErrorResponse & {
  readonly statusCode?: number;
}): string => {
  switch (code) {
    case 400:
      return `Invalid Subscription Id|${code}|${message}`;
    case 404:
      return `Subscription not found|${code}|${message}`;
    default:
      return `APIM Generic error|${code || "no-code-returned"}|${message}`;
  }
};
