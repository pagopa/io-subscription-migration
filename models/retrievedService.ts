/* eslint-disable prettier/prettier */
import {
  NonEmptyString,
  OrganizationFiscalCode,
} from "@pagopa/ts-commons/lib/strings";
import * as t from "io-ts";

const RetrievedServiceDocumentR = t.interface({
  organizationFiscalCode: OrganizationFiscalCode,
  subscriptionId: NonEmptyString,
});
const RetrievedServiceDocumentO = t.partial({
  serviceName: NonEmptyString,
});

export const RetrievedServiceDocument = t.exact(
  t.intersection(
    [RetrievedServiceDocumentR, RetrievedServiceDocumentO],
    "RetrievedServiceDocument"
  )
);
export type RetrievedServiceDocument = t.TypeOf<
  typeof RetrievedServiceDocument
>;
