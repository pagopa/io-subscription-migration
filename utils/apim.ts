import { ApiManagementClient } from "@azure/arm-apimanagement";
import { DefaultAzureCredential } from "@azure/identity";

export interface IServicePrincipalCreds {
  readonly clientId: string;
  readonly secret: string;
  readonly tenantId: string;
}

export interface IAzureApimConfig {
  readonly subscriptionId: string;
  readonly apimResourceGroup: string;
  readonly apim: string;
}
// eslint-disable-next-line prefer-arrow/prefer-arrow-functions
export function getApiClient(subscriptionId: string): ApiManagementClient {
  return new ApiManagementClient(new DefaultAzureCredential(), subscriptionId);
}
