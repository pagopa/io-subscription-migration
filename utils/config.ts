/**
 * Config module
 *
 * Single point of access for the application confguration. Handles validation on required environment variables.
 * The configuration is evaluate eagerly at the first access to the module. The module exposes convenient methods to access such value.
 */

import * as t from "io-ts";

import * as E from "fp-ts/lib/Either";
import { pipe } from "fp-ts/lib/function";

import { readableReport } from "@pagopa/ts-commons/lib/reporters";
import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
import { withDefault } from "@pagopa/ts-commons/lib/types";
import {
  IntegerFromString,
  NumberFromString
} from "@pagopa/ts-commons/lib/numbers";

// Environment configuration to connect to IO CosmosDB
//   needed in order to fetch changes on Services collection
export type IDecodableConfigCosmosDB = t.TypeOf<
  typeof IDecodableConfigCosmosDB
>;
export const IDecodableConfigCosmosDB = t.interface({
  COSMOSDB_CONNECTIONSTRING: NonEmptyString,
  COSMOSDB_KEY: NonEmptyString,
  COSMOSDB_NAME: NonEmptyString,
  COSMOSDB_SERVICES_COLLECTION: NonEmptyString,
  COSMOSDB_SERVICES_LEASE_COLLECTION: NonEmptyString,
  COSMOSDB_URI: NonEmptyString
});

// Environment configuration to connect to IO APIM instance
//   needed in order to query API Subscription Keys
export type IDecodableConfigAPIM = t.TypeOf<typeof IDecodableConfigAPIM>;
export const IDecodableConfigAPIM = t.interface({
  APIM_CLIENT_ID: NonEmptyString,
  APIM_RESOURCE_GROUP: NonEmptyString,
  APIM_SECRET: NonEmptyString,
  APIM_SERVICE_NAME: NonEmptyString,
  APIM_SUBSCRIPTION_ID: NonEmptyString,
  APIM_TENANT_ID: NonEmptyString
});

// Environment configuration to connect to dedicate db instance
//   needed in order to persist migration data
export type IDecodableConfigPostgreSQL = t.TypeOf<
  typeof IDecodableConfigPostgreSQL
>;
export const IDecodableConfigPostgreSQL = t.interface({
  DB_HOST: NonEmptyString,
  DB_IDLE_TIMEOUT: withDefault(NumberFromString, 30000),
  DB_NAME: NonEmptyString,
  DB_PASSWORD: NonEmptyString,
  DB_PORT: NumberFromString,
  DB_SCHEMA: NonEmptyString,
  DB_TABLE: NonEmptyString,
  DB_USER: NonEmptyString
});

// Environment configuration to connect to Application Insight instance
//   needed in order to monitoring basic events and custom events
export type IDecodableConfigAppInsight = t.TypeOf<
  typeof IDecodableConfigAppInsight
>;
export const IDecodableConfigAppInsight = t.intersection([
  t.interface({
    APPINSIGHTS_INSTRUMENTATIONKEY: NonEmptyString
  }),
  t.partial({
    APPINSIGHTS_DISABLE: NonEmptyString,
    APPINSIGHTS_SAMPLING_PERCENTAGE: withDefault(IntegerFromString, 5)
  })
]);

export type IConfig = t.TypeOf<typeof IConfig>;
export const IConfig = t.intersection([
  IDecodableConfigCosmosDB,
  IDecodableConfigAPIM,
  IDecodableConfigPostgreSQL,
  IDecodableConfigAppInsight,
  t.interface({
    // default function app storage connection
    AzureWebJobsStorage: NonEmptyString,

    INTERNAL_STORAGE_CONNECTION_STRING: NonEmptyString,
    QUEUE_ADD_SERVICE_TO_MIGRATIONS: NonEmptyString,

    isProduction: t.boolean
  })
]);

export const envConfig = {
  ...process.env,
  isProduction: process.env.NODE_ENV === "production"
};

// No need to re-evaluate this object for each call
const errorOrConfig: t.Validation<IConfig> = IConfig.decode(envConfig);

/**
 * Read the application configuration and check for invalid values.
 * Configuration is eagerly evalued when the application starts.
 *
 * @returns either the configuration values or a list of validation errors
 */
export const getConfig = (): t.Validation<IConfig> => errorOrConfig;

/**
 * Read the application configuration and check for invalid values.
 * If the application is not valid, raises an exception.
 *
 * @returns the configuration values
 * @throws validation errors found while parsing the application configuration
 */
export const getConfigOrThrow = (): IConfig =>
  pipe(
    errorOrConfig,
    E.getOrElseW((errors: ReadonlyArray<t.ValidationError>) => {
      throw new Error(`Invalid configuration: ${readableReport(errors)}`);
    })
  );
