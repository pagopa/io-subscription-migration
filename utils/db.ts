import { pipe } from "fp-ts/lib/function";
import { DatabaseError, Pool, QueryResult } from "pg";
import * as TE from "fp-ts/TaskEither";
import * as t from "io-ts";
import { IDecodableConfigPostgreSQL } from "./config";

// eslint-disable-next-line functional/no-let
let singletonPool: Pool;
export const getPool = (config: IDecodableConfigPostgreSQL): Pool => {
  if (!singletonPool) {
    singletonPool = new Pool({
      database: config.DB_NAME,
      host: config.DB_HOST,
      idleTimeoutMillis: config.DB_IDLE_TIMEOUT,
      max: 20,
      password: config.DB_PASSWORD,
      port: config.DB_PORT,
      ssl: true,
      user: config.DB_USER
    });
  }
  return singletonPool;
};

export const queryDataTable = (
  pool: Pool,
  query: string
): TE.TaskEither<DatabaseError, QueryResult> =>
  pipe(
    TE.tryCatch(
      () => pool.query(query),
      error => error as DatabaseError
    ),
    TE.mapLeft(e => e)
  );

export const ResultRow = t.interface({
  count: t.string,
  status: t.string
});
export type ResultRow = t.TypeOf<typeof ResultRow>;

export const ResultSet = t.interface({
  rowCount: t.number,
  rows: t.readonlyArray(ResultRow)
});
export type ResultSet = t.TypeOf<typeof ResultSet>;

export default getPool;
