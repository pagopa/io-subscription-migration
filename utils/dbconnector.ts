import { Pool } from "pg";
import { IDecodableConfigPostgreSQL } from "./config";

// eslint-disable-next-line functional/no-let
let singletonPool: Pool;
export const getPool = (config: IDecodableConfigPostgreSQL): Pool => {
  if (!singletonPool) {
    singletonPool = new Pool({
      database: config.DB_NAME,
      host: config.DB_HOST,
      idleTimeoutMillis: config.DB_IDLE_TIMEOUT,
      max: 20,
      password: config.DB_PASSWORD,
      port: config.DB_PORT,
      ssl: true,
      user: config.DB_USER
    });
  }
  return singletonPool;
};

export default getPool;
