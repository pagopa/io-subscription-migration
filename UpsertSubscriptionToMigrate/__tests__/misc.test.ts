import * as E from "fp-ts/lib/Either";
import * as O from "fp-ts/lib/Option";
import {
  NonEmptyString,
  OrganizationFiscalCode
} from "@pagopa/ts-commons/lib/strings";
import {
  createUpsertSql,
  parseOwnerIdFullPath,
  validateDocument
} from "../handler";
import { IDecodableConfigPostgreSQL } from "../../utils/config";
import { MigrationRowDataTable } from "../../models/Domain";
import {
  ApimDelegateUserResponse,
  ApimOrganizationUserResponse,
  ApimUserResponse
} from "../../models/DomainApimResponse";
import { isRight } from "fp-ts/lib/Either";
import {
  RetrievedService,
  toAuthorizedCIDRs,
  toAuthorizedRecipients
} from "@pagopa/io-functions-commons/dist/src/models/service";
import {
  IWithinRangeIntegerTag,
  NonNegativeInteger
} from "@pagopa/ts-commons/lib/numbers";

describe("validate Document", () => {
  it("should validate a valid document", () => {
    const doc: RetrievedService = {
      authorizedCIDRs: toAuthorizedCIDRs([]),
      authorizedRecipients: toAuthorizedRecipients([]),
      departmentName: "MyDeptName" as NonEmptyString,
      isVisible: true,
      maxAllowedPaymentAmount: 0 as
        | 9999999999
        | (number & IWithinRangeIntegerTag<0, 9999999999>),
      organizationFiscalCode: "00000000000" as OrganizationFiscalCode,
      organizationName: "MyOrgName" as NonEmptyString,
      requireSecureChannels: false,
      serviceId: "MySubscriptionId" as NonEmptyString,
      serviceName: "MyServiceName" as NonEmptyString,
      id: "123" as NonEmptyString,
      version: 1 as NonNegativeInteger,
      kind: "IRetrievedService",
      _rid: "rid",
      _ts: 1,
      _self: "self",
      _etag: "etag"
    };
    const res = validateDocument(doc);
    expect(E.isRight(res)).toBe(true);
  });
});
describe("parseOwnerIdFullPath", () => {
  it("should return None for an empyy path", async () => {
    const fullPath = "" as NonEmptyString;
    const parsed = parseOwnerIdFullPath(fullPath);
    expect(O.isNone(parsed)).toBe(true);
  });
  it("should return None for an invalid path", async () => {
    const fullPath = "This\\IsAnInvalid\\Path" as NonEmptyString;
    const parsed = parseOwnerIdFullPath(fullPath);
    expect(O.isNone(parsed)).toBe(true);
  });
  it("should return None for a malformed path", async () => {
    const fullPath = "/subscriptions/subid/resourceGroups/providers/Microsoft.ApiManagement/service/users/5931a75ae4bbd512a88c680b" as NonEmptyString;
    const parsed = parseOwnerIdFullPath(fullPath);
    expect(O.isNone(parsed)).toBe(true);
  });
  it("should return Some for a valid path", async () => {
    const fullPath = "/subscriptions/subid/resourceGroups/resourceGroupName/providers/Microsoft.ApiManagement/service/apimServiceName/users/5931a75ae4bbd512a88c680b" as NonEmptyString;
    const expected = "5931a75ae4bbd512a88c680b";
    const parsed = parseOwnerIdFullPath(fullPath);
    expect(O.isSome(parsed)).toBe(true);
    if (O.isSome(parsed)) {
      expect(parsed.value).toBe(expected);
    } else {
      throw new Error("Expected some value, received other");
    }
  });
});

describe("ApimOrganizationUserResponse", () => {
  it("should validate a valid Organization User", () => {
    const value = {
      email: "email@test.com",
      firstName: "TestNome",
      id: "123",
      lastName: "TestCognome",
      note: "01234567891"
    };

    expect(isRight(ApimOrganizationUserResponse.decode(value))).toBe(true);
    expect(ApimOrganizationUserResponse.is(value)).toBe(true);
  });
  it("should not validate as valid Delegate User", () => {
    const value = {
      email: "email@test.com",
      firstName: "TestNome",
      id: "123",
      lastName: "TestCognome",
      note: "01234567891"
    };
    expect(isRight(ApimDelegateUserResponse.decode(value))).toBe(false);
    expect(ApimDelegateUserResponse.is(value)).toBe(false);
  });
});

describe("ApimDelegateUserResponse", () => {
  it("should validate a valid Delegate User", () => {
    const value = {
      id:
        "/subscriptions/subid/resourceGroups/resourceGroupName/providers/Microsoft.ApiManagement/service/apimServiceName/users/01EYNPZXQJF9A2DBTH5GYB951V",
      firstName: "Nome",
      lastName: "Cognome",
      email: "email@test.com",
      note: ""
    };

    expect(isRight(ApimDelegateUserResponse.decode(value))).toBe(true);
    expect(ApimDelegateUserResponse.is(value)).toBe(true);
  });
  it("sould not validate as a valide Organization", () => {
    const value = {
      id:
        "/subscriptions/subid/resourceGroups/resourceGroupName/providers/Microsoft.ApiManagement/service/apimServiceName/users/01EYNPZXQJF9A2DBTH5GYB951V",
      firstName: "Nome",
      lastName: "Cognome",
      email: "email@test.com",
      note: ""
    };
    expect(isRight(ApimOrganizationUserResponse.decode(value))).toBe(false);
    expect(ApimOrganizationUserResponse.is(value)).toBe(false);
  });
});

describe("ApimUserResponse", () => {
  it("should validate a valid User", () => {
    const value = {
      email: "email@test.com",
      firstName: "TestNome",
      id: "123",
      lastName: "TestCognome"
    };
    const res = ApimUserResponse.decode(value);
    expect(isRight(res)).toBe(true);
  });
  it("should validate an organization", () => {
    const value = {
      email: "email@test.com",
      firstName: "TestNome",
      id: "123",
      lastName: "TestCognome",
      note: "01234567891"
    };
    const res = ApimUserResponse.decode(value);
    expect(isRight(res)).toBe(true);
  });
});

describe("createUpsertSql", () => {
  it("should compose correct upsert sql", async () => {
    const config = {
      DB_SCHEMA: "SelfcareIOSubscriptionMigrations",
      DB_TABLE: "migrations"
    } as IDecodableConfigPostgreSQL;
    const data = ({
      subscriptionId: "subId1",
      organizationFiscalCode: "12345678901",
      sourceId: "01EYNPZXQJF9A2DBTH5GYB951V",
      sourceName: "source name",
      sourceSurname: "source surname",
      sourceEmail: "source email",
      serviceVersion: 0,
      serviceName: "Service Test"
    } as unknown) as MigrationRowDataTable;
    const expected = `insert into "SelfcareIOSubscriptionMigrations"."migrations" ("organizationFiscalCode", "serviceName", "serviceVersion", "sourceEmail", "sourceId", "sourceName", "sourceSurname", "subscriptionId") values ('12345678901', 'Service Test', 0, 'source email', '01EYNPZXQJF9A2DBTH5GYB951V', 'source name', 'source surname', 'subId1') on conflict ("subscriptionId") do update set "organizationFiscalCode" = excluded."organizationFiscalCode", "serviceVersion" = excluded."serviceVersion", "serviceName" = excluded."serviceName" where "migrations"."status" < 'PENDING' and "migrations"."serviceVersion" < excluded."serviceVersion"`;

    const sql = createUpsertSql(config)(data);
    expect(sql.trim()).toBe(expected.trim());
  });

  it("should escape single quotes", async () => {
    const config = {
      DB_SCHEMA: "SelfcareIOSubscriptionMigrations",
      DB_TABLE: "migrations"
    } as IDecodableConfigPostgreSQL;
    const data = ({
      subscriptionId: "subId2",
      organizationFiscalCode: "12345678901",
      sourceId: "01EYNPZXQJF9A2DBTH5GYB951V",
      sourceName: "source name",
      sourceSurname: "source surname",
      sourceEmail: "l'email", // <-- single quote!
      serviceVersion: 0,
      serviceName: "Service Test"
    } as unknown) as MigrationRowDataTable;
    const expected = `insert into "SelfcareIOSubscriptionMigrations"."migrations" ("organizationFiscalCode", "serviceName", "serviceVersion", "sourceEmail", "sourceId", "sourceName", "sourceSurname", "subscriptionId") values ('12345678901', 'Service Test', 0, 'l''email', '01EYNPZXQJF9A2DBTH5GYB951V', 'source name', 'source surname', 'subId2') on conflict ("subscriptionId") do update set "organizationFiscalCode" = excluded."organizationFiscalCode", "serviceVersion" = excluded."serviceVersion", "serviceName" = excluded."serviceName" where "migrations"."status" < 'PENDING' and "migrations"."serviceVersion" < excluded."serviceVersion"`;

    const sql = createUpsertSql(config)(data);
    expect(sql.trim()).toBe(expected.trim());
  });
});
