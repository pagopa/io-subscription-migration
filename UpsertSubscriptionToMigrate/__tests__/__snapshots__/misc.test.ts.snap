// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createUpsertSql should compose correct upsert sql 1`] = `"insert into \\"SelfcareIOSubscriptionMigrations\\".\\"migrations\\" (\\"hasBeenVisibleOnce\\", \\"isVisible\\", \\"organizationFiscalCode\\", \\"serviceName\\", \\"serviceVersion\\", \\"sourceEmail\\", \\"sourceId\\", \\"sourceName\\", \\"sourceSurname\\", \\"subscriptionId\\") values (true, true, '12345678901', 'Service Test', 0, 'source email', '01EYNPZXQJF9A2DBTH5GYB951V', 'source name', 'source surname', 'subId1') on conflict (\\"subscriptionId\\") do update set \\"hasBeenVisibleOnce\\" = \\"hasBeenVisibleOnce\\" OR excluded.\\"hasBeenVisibleOnce\\",\\"isVisible\\" = excluded.\\"isVisible\\",\\"organizationFiscalCode\\" = excluded.\\"organizationFiscalCode\\",\\"serviceName\\" = excluded.\\"serviceName\\",\\"serviceVersion\\" = excluded.\\"serviceVersion\\" where \\"migrations\\".\\"status\\" < 'PENDING' and \\"migrations\\".\\"serviceVersion\\" <= excluded.\\"serviceVersion\\""`;

exports[`createUpsertSql should escape single quotes 1`] = `"insert into \\"SelfcareIOSubscriptionMigrations\\".\\"migrations\\" (\\"hasBeenVisibleOnce\\", \\"isVisible\\", \\"organizationFiscalCode\\", \\"serviceName\\", \\"serviceVersion\\", \\"sourceEmail\\", \\"sourceId\\", \\"sourceName\\", \\"sourceSurname\\", \\"subscriptionId\\") values (true, true, '12345678901', 'Service Test', 0, 'l''email', '01EYNPZXQJF9A2DBTH5GYB951V', 'source name', 'source surname', 'subId2') on conflict (\\"subscriptionId\\") do update set \\"hasBeenVisibleOnce\\" = \\"hasBeenVisibleOnce\\" OR excluded.\\"hasBeenVisibleOnce\\",\\"isVisible\\" = excluded.\\"isVisible\\",\\"organizationFiscalCode\\" = excluded.\\"organizationFiscalCode\\",\\"serviceName\\" = excluded.\\"serviceName\\",\\"serviceVersion\\" = excluded.\\"serviceVersion\\" where \\"migrations\\".\\"status\\" < 'PENDING' and \\"migrations\\".\\"serviceVersion\\" <= excluded.\\"serviceVersion\\""`;
